### PR TITLE
Integrate TextFormation

### PR DESCRIPTION
**Pull request by thecoolwinter**
# Description

This PR integrates the `TextFormation` library from the awesome @mattmassicotte and @ChimeHQ. It implements the necessary type conformances for `STTextView`, and adds the correct delegate methods in `STTextViewController` to handle edits.

Details:
- Added a `textFilters` array and `setUpTextFormation()` method to `STTextViewController` to set up filters and store them on the class.
- Added the `textView(_ textView: STTextView, shouldChangeTextIn affectedCharRange: NSTextRange, replacementString: String?)` method to `STTextViewController` to apply edits through `TextFormation`
- Created three filters
    - `NewlineFilter`: This filter is a near 1:1 copy of TextFormation's newline filter. This may be removed in the future, but TextFormation's default implementation caused every other new line to be broken.
    - `DeleteWhitespaceFilter`: Deletes segments of whitespace, eg if I have my whitespace set to 4 spaces, when I try to delete an empty line with 8 spaces it will delete the first 4 instead of just one space.
    - `TabReplacementFilter`: Replaces tab characters with the given indentation unit (this can be a tab character, but this is a point at which to handle spaces or any other alternatives).
- Removed the `keyDown` implementation on `STTextViewController`. This ended up causing problems, and `STTextView` handles tab insertion fine without it. In the future this should be removed, but it seems that `STTextViewController` loses focus on tab insertion without it. So, it was left in temporarily.

# Related Issues

- #79 
- #80 
- #78 
- #23 
- #37 

# Screenshots

Newlines:

https://user-images.githubusercontent.com/35942988/215297117-4eb68a67-9059-4022-bd6a-cd9f4ca1b437.mov

Bracket Pairs:

https://user-images.githubusercontent.com/35942988/215297139-42b61021-78b7-4426-8355-6dd95a21ef7b.mov

Tabulation:

https://user-images.githubusercontent.com/35942988/215297144-2d8502ec-e239-4632-8dde-03fd8ff20bdb.mov

*^ It's not clear but when it only goes left by a couple characters I'm using my arrow keys not the delete button.*

### CODE REVIEW

**Comment by thecoolwinter** [comment]
I'm not linking to #80 yet because that issue I think also covers the need to let users use a tab character. That's outside the scope of this PR though.

File: Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift
Comment (Wouter01) [inline Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift:23]: Small thing, but wouldn't it be clearer if the `self.selectedRange()` method was renamed to `self.setSelectedRange()`? I was confused, because at first sight it looks like the computed var calls itself.

**Review by Wouter01 — APPROVED** [review]
Awesome work!

File: Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift
Comment (thecoolwinter) [reply Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift:23]: That variable is part of the conformance to `TextInterface` so it can't be changed sadly.

Comment (thecoolwinter) [reply Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift:23]: I believe that's implemented by STTextView so it's outside our implementation. Let me check

Comment (thecoolwinter) [reply Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift:23]: Yeah that can't be changed.

**Comment by thecoolwinter** [comment]
I'll wait for one more review before merging since this is a large change.

File: Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift
Comment (Wouter01) [reply Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift:23]: > Yeah that can't be changed.

Oh okay, my bad!

**Comment by austincondiff** [comment]
Way to go @thecoolwinter, great work!

**Comment by mattmassicotte** [comment]
@thecoolwinter wow awesome!

I'm very interesting in investigating what's up with `NewlineFilter`. Almost sounds like a line endings issue.

**Comment by matthijseikelenboom** [comment]
@thecoolwinter Very nice work man! Only comment that I have is that the order of struct properties and functions is a bit all over the place.

**Comment by thecoolwinter** [comment]
> @thecoolwinter wow awesome!
> 
> 
> 
> I'm very interested in investigating what's up with `NewlineFilter`. Almost sounds like a line endings issue.

Yes and I was going to open an issue on TextFormation. For some reason it's recognizing the beginning of the previous line as the end of the line, but only every other newline. I found that removing the line ending logic fixed it temporarily.

**Comment by thecoolwinter** [comment]
> @thecoolwinter Very nice work man! Only comment that I have is that the order of struct properties and functions is a bit all over the place.

Yeah you're right. I can clean that up a little more real quick

File: Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift
Comment (mattmassicotte) [inline Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift:43]: Are you not able to use `textContentStorage. applyMutation ` here?

Comment (thecoolwinter) [reply Sources/CodeEditTextView/Extensions/STTextView+/STTextView+TextInterface.swift:43]: Oh shoot I didn't realize it was available.

**Review by lukepistrol — APPROVED** [review]
Looks nice! 👍

Kudos also to @mattmassicotte for his work on the whole Chime suite 😉

**Comment by mattmassicotte** [comment]
> Looks nice! 👍
> 
> Kudos also to @mattmassicotte for his work on the whole Chime suite 😉

I'm so glad you are finding all of this useful. Something that could be interesting is turning the STTextView + TextFormation integration itself into a package. Just food for thought.

**Pull request merged**
